### PR TITLE
Delegate

### PR DESCRIPTION
**Pull request by bitpshr**
This PR includes a delegate API without using extension events, as per #11. A basic set of unit tests is also included.

### CODE REVIEW

File: tests/unit/delegate.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import assert = require('intern/chai!assert');
+import { emit } from 'dojo-core/on';
+import { Handle } from 'dojo-core/interfaces';
+import delegate from 'src/delegate';
+import * as dom from 'src/dom';
+
+let container = document.createElement('div');
+let handles: (Handle)[] = [];
+
+document.body.appendChild(container);
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: This should be in a `setup` function in the test suite, and there should be `teardown` logic to remove it.

File: src/delegate.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import { Handle } from 'dojo-core/interfaces';
+import on, { ExtensionEvent } from 'dojo-core/on';
+
+export default function delegate(target: HTMLElement, selector: string, type: string, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: ExtensionEvent, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: (string | ExtensionEvent)[], listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: any, listener: (event: UIEvent) => void): Handle {
+	function matches(target: EventTarget) {
+		let node = <any> target;
```
Comment (kitsonk) [inline src/delegate.ts]: Why the any cast?  IMO, we should at least comment whenever we do an any cast, because elsewise it looks like we are simply "cheating" to get around strong typing, and we shouldn't do that without explaining ourselves.

File: tests/unit/delegate.ts (hunk 1 above)
Comment (kitsonk) [inline tests/unit/delegate.ts]: Argreed, because you might be surprised when some other test has already wiped your object off the DOM for you!

**Comment by kitsonk** [comment]
On the overall test coverage, I can't do the mental gymnastics enough to see if every overloaded call to `delegate` is covered.

File: src/delegate.ts (hunk 2 above)
Comment (kfranqueiro) [inline src/delegate.ts]: Unfortunately, for some reason, TypeScript's typings include `webkitMatchesSelector` and `msMatchesSelector` in `Element`, but not `matches` (the actual standard).  So at the very least, `<any>` needs to be used for that.  However, in this location it's actually hiding a number of other issues, including the fact that this function is returning either a boolean or a Node, which seems pretty awkward.

I came up with the following instead:

```ts
	function matches(target: HTMLElement) {
		let node = target;
		let matchMethod = (function() {
			// TS typings include the vendor-prefixed methods used here, but not the standard
			if (typeof (<any> node).matches === 'function') {
				return 'matches';
			}
			if (typeof node.webkitMatchesSelector === 'function') {
				return 'webkitMatchesSelector';
			}
			if (typeof node.msMatchesSelector === 'function') {
				return 'msMatchesSelector';
			}
		})();

		// TS7017
		while (!(<any> node)[matchMethod](selector)) {
			if (!node.parentNode || !(<any> node.parentNode)[matchMethod]) {
				return null;
			}
			node = <HTMLElement> node.parentNode;
		}

		return node;
	}

	return on(<EventTarget> target, type, function (event: Event) {
		let matchedEventTarget = matches(<HTMLElement> event.target);
		if (matchedEventTarget) {
			return listener.call(matchedEventTarget, event);
		}
	});
```

File: src/delegate.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		let matchMethod = (function() {
```
Comment (kfranqueiro) [inline src/delegate.ts]: We shouldn't have to execute this logic every single time each delegated event handler runs.  This looks like a prime candidate for a `has` feature test to me (which should go in `src/has` which doesn't exist yet).  It should be fine for the feature test to return the method name.

File: tests/unit/delegate.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+		container.appendChild(buttonOne);
+		container.appendChild(buttonTwo);
+
+		buttonOne.click();
+		buttonTwo.click();
+		assert.strictEqual(called, 2);
+	},
+
+	'CSS selector and text node target'() {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Not really sure why this says "text node target" since AFAICT we're emitting off of a div element?

File: src/delegate.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+			}
+			node = node.parentNode;
+		}
+
+		return node;
+	}
+
+	return on(<EventTarget> target, type, function (event: Event) {
+		let matchedEventTarget = matches(event.target);
+		if (matchedEventTarget) {
```
Comment (kfranqueiro) [inline src/delegate.ts]: Unfortunately, `matches` is subject to the same undesirable behavior as `querySelector` itself, which we probably want to avoid tripping on.

Test case:
1. Create a 3-level element structure, giving the top element a `className` of `ancestor` and the bottom element a `className` of `child`
2. Set up a delegated event handler on the *middle* element, using the selector `.ancestor .child`
3. You would normally expect this *not* to match, given that `ancestor` is only on an element *above* the given root, but it *will* match.

@neonstalwart suggested adding a `contains` check to avoid this (i.e. `if (matchedEventTarget && dom.contains(target, matchedEventTarget))`).

We should also add a test for this case based on the above steps.

(Note: I have to fix a bug in `dom.contains` for this to work in all cases; I should have that fixed sometime this morning.)

Comment (kfranqueiro) [inline src/delegate.ts]: FYI, `dom.contains` is now fixed in master (see #18).

**Comment by kfranqueiro** [comment]
RE @kitsonk's comment about what overloads we're testing, IIRC I didn't see any tests sending arrays of event types, though it's ultimately just forwarding that argument straight through to `on` anyway.

As for extension events, we're still determining whether anything actually warrants being an extension event, or if we (a) don't need them or (b) can accommodate what they accomplish by other means.

File: tests/unit/delegate.ts (hunk 4 above)
Comment (bitpshr) [inline tests/unit/delegate.ts]: The delegation selector targets a div element, but the actual click event as emitted on a text node. I updated the test name to be a bit more clear, maybe we can still do better...

**Comment by kfranqueiro** [comment]
RE https://github.com/dojo/dom/pull/16#discussion_r31141294, @bitpshr correctly pointed out that `contains` can't actually solve this problem, since it's the ancestor, not the child, that would need to be validated.

The way that Dojo 1 solves this problem (an idea it supposedly borrowed from Sizzle, though I haven't yet verified its origin) is to give the target node a temporary ID if it doesn't already have one, and prepend that to all selectors that are passed.  The implementation of that is [here](https://github.com/dojo/dojo/blob/1.10.4/selector/lite.js#L97-L129) and it's used [here](https://github.com/dojo/dojo/blob/1.10.4/selector/lite.js#L274).

However, I'm a little wary of how hard to follow some of that code is (especially the [RegExp that goes hunting for selectors between commas](https://github.com/dojo/dojo/blob/1.10.4/selector/lite.js#L7) - how naive would simply splitting on comma be instead?), so it's not very attractive from a maintenance perspective.

Ben and myself have been arguing both sides of this - any time where this behavior bites is likely resolvable by tightening the selector that is passed, but Bryan and I are concerned that it's likely to catch a lot of people off-guard.  I can fathom leaving it out for now, given that we might not merge this immediately anyway, but it definitely warrants more investigation because I think it's going to end up being worth it to end users.

**Comment by bitpshr** [comment]
@kfranqueiro I added logic to handle rooted match queries. `delegate` now mandates that matched elements are descendants of the delegated target. For now, I don't think this needs to be a separate method. We were also able to nix the `relativeHierarchicalSelector` stuff from other implementations.

File: src/delegate.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ * @example
+ * dom.delegate(document.body, 'li', ['click', 'mouseover'], function () {
+ *     // ...
+ * });
+ */
+export default function delegate(target: HTMLElement, selector: string, type: string, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: ExtensionEvent, listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: (string | ExtensionEvent)[], listener: (event: UIEvent) => void): Handle;
+export default function delegate(target: HTMLElement, selector: string, type: any, listener: (event: UIEvent) => void): Handle {
+	function matches(node: HTMLElement, selector: string) {
```
Comment (kfranqueiro) [inline src/delegate.ts]: I am not sure that this function even needs to exist.  Is this a remnant of an earlier approach?  AFAICT directly calling the matches method down below where this is currently called works fine.  (I've changed it locally and the unit tests still pass.)

File: src/delegate.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		}
+
+		let selectors = selector.split(',');
+		for (let i = 0; i < selectors.length; i++) {
+			selectors[i] = '[id="' + id + '"] ' + selectors[i];
+		}
+		let tempSelector = selectors.join(',');
+		let matchedEventTarget = matches(<HTMLElement> event.target, tempSelector);
+
+		try {
```
Comment (kfranqueiro) [inline src/delegate.ts]: The try/finally should be around the `matches` call (as it was in Dojo 1), not around the listener invocation.  (I've fixed this on my own branch.)

File: src/has.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { add } from 'dojo-core/has';
+
+add('element-match', function () {
```
Comment (kfranqueiro) [inline src/has.ts]: FYI, I'm renaming this to `dom-element-matches`.

File: tests/unit/delegate.ts [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+	setup() {
+		document.body.appendChild(container);
+	},
+
+	teardown() {
+		document.body.removeChild(container);
+	},
+
+	afterEach() {
+		while (handles.length > 0) {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Nit-picking, but this can use for...of to avoid manipulating the array every time (then just reset handles to a new array).  (I'm fixing this on my branch.)

File: tests/unit/delegate.ts [hunk 10]
```diff
@@ -0,0 +1,180 @@
+import registerSuite = require('intern!object');
+import assert = require('intern/chai!assert');
+import { emit } from 'dojo-core/on';
+import { Handle } from 'dojo-core/interfaces';
+import delegate from 'src/delegate';
+import * as dom from 'src/dom';
+
+let container = document.createElement('div');
+let handles: (Handle)[] = [];
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Not sure why there were parens around `Handle` here.

File: tests/unit/delegate.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+		buttonOne.click();
+		buttonTwo.click();
+		assert.strictEqual(called, 2);
+	},
+
+	'CSS selector with array of events'() {
+		let called = 0;
+		let button = document.createElement('button');
+
+		delegate(container, 'button', ['click'], function() {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Missing spaces inside brackets.

File: tests/unit/delegate.ts [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+import assert = require('intern/chai!assert');
+import { emit } from 'dojo-core/on';
+import { Handle } from 'dojo-core/interfaces';
+import delegate from 'src/delegate';
+import * as dom from 'src/dom';
+
+let container = document.createElement('div');
+let handles: (Handle)[] = [];
+
+function testDelegate(...args: any[]) {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: About 5 tests weren't calling this function (they were calling `delegate` directly and thus their handles were never cleaned up), but it's also trivial to wrap this around each delegate call (i.e. `handles.push(delegate(...))`) so I'll be doing that instead.

File: tests/unit/delegate.ts [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+			called++;
+		});
+
+		container.appendChild(button);
+
+		button.click();
+		assert.strictEqual(called, 1);
+	},
+
+	'CSS selector and text node event target'() {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: I think I commented on this the other day and you answered, but I'm still confused.  This test is registering a delegated handler on the container, listening for child div clicks, and is emitting a click event on a child div.  There's nothing in here dealing with text nodes, so I don't understand why the title says otherwise.

File: tests/unit/delegate.ts [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+
+	'Listening on body'() {
+		let called = 0;
+		let button = document.createElement('button');
+
+		testDelegate(document.body, 'button', 'click', function() {
+			called++;
+		});
+
+		document.body.appendChild(button);
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: This isn't getting cleaned up.  I'm changing it to add to `container` since that should still effectively test it.

File: tests/unit/delegate.ts [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+
+	'Should not throw when selector is not found'() {
+		let called = 0;
+		let button = document.createElement('button');
+
+		testDelegate(container, 'not-found', 'click', function() {
+			called++;
+		});
+
+		document.body.appendChild(button);
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: This isn't getting cleaned up either, but is also completely unnecessary, so I'll simplify it.

File: tests/unit/delegate.ts [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+		let child = document.createElement('button');
+
+		testDelegate(parent, 'div button', 'click', function() {
+			called++;
+		});
+
+		parent.appendChild(child);
+		container.appendChild(parent);
+
+		child.click()
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Missing semicolon.

File: tests/unit/delegate.ts (hunk 13 above)
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Discussed with Paul - this was intended to test emitting on a text node.  However, text nodes don't have `.matches`.  If Dojo 1 was handling this, we'll have to investigate how (bearing in mind that it relied on `dojo/query` for `matches`).

File: src/delegate.ts (hunk 6 above)
Comment (bitpshr) [inline src/delegate.ts]: `Element.matches` determines if a single node matches a single selector. In this case, I believe we need to traverse up if this element is not matched. We also should be special-casing Text nodes.

File: tests/unit/delegate.ts (hunk 13 above)
Comment (bitpshr) [inline tests/unit/delegate.ts]: This test should be emitting the event of `div.firstChild`. This may highlight a bug about Text nodes.

File: tests/unit/delegate.ts (hunk 15 above)
Comment (bitpshr) [inline tests/unit/delegate.ts]: Should not *fire.

File: src/delegate.ts (hunk 6 above)
Comment (kfranqueiro) [inline src/delegate.ts]: This originated from https://github.com/dojo/dojo/blob/1.10.4/on.js#L219-L223.  Paul makes a valid point that we would need to traverse upwards to match the selector on parents (in order to honor events that don't originate from a matching element, but would bubble up to one).

File: tests/unit/delegate.ts [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+		}
+		container.innerHTML = '';
+	},
+
+	'CSS selector matching multiple elements'() {
+		let called = 0;
+		let buttonOne = document.createElement('button');
+		let buttonTwo = document.createElement('button');
+
+		delegate(container, 'button', 'click', function() {
```
Comment (kfranqueiro) [inline tests/unit/delegate.ts]: Anonymous functions should have a space between `function` and `(`.

File: src/delegate.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+			if (!node.parentNode || !(<any> node.parentNode)[matchMethod]) {
+				return null;
+			}
+			node = <HTMLElement> node.parentNode;
+		}
+
+		return node;
+	}
+
+	return on(<EventTarget> target, type, function(event: Event) {
```
Comment (kfranqueiro) [inline src/delegate.ts]: The cast to EventTarget here is unnecessary.  HTMLElement -> Element -> Node -> EventTarget.

**Comment by kfranqueiro** [comment]
Cleaned-up branch at https://github.com/kfranqueiro/dom/commits/delegate.  Not merging right now, pending more events API exploration.

**Comment by bitpshr** [comment]
@kfranqueiro let me know if you'd like me to address any of this or take it further.
